source-oracle/replication.go

```diff
@@ -0,0 +1,899 @@
+			"startSCN":  msg.startSCN,
+			"endSCN":    msg.endSCN,
+			"ssn":       msg.ssn,
+			"rsid":      msg.rsID,
+			"csf":       msg.csf,
+		}).Trace("received message")
+
+		// last message indicated a continuation of SQL_REDO and SQL_UNDO values
+		// so this row's sqls will be appended to the last
+		if lastMsg != nil && lastMsg.csf == 1 {
+			// sanity check the (SSN, RSID) tuple matches between the rows
+			if lastMsg.ssn != msg.ssn || lastMsg.rsID != msg.rsID {
+				return nil, fmt.Errorf("expected SSN and RSID of continued rows to match: (%d, %s) != (%d, %s)", lastMsg.ssn, lastMsg.rsID, msg.ssn, msg.rsID)
+			}
+			lastMsg.sql += msg.sql
+			lastMsg.undoSql += msg.undoSql
+			lastMsg.csf = msg.csf
+			continue
+		}
+
+		// If this change event is on a table we're not capturing, skip doing any
+		// further processing on it.
+		var streamID = sqlcapture.JoinStreamID(msg.owner, msg.tableName)
+		if !s.tableActive(streamID) {
+			continue
+		}
+
+		// If logminer can't find the dictionary for a SQL statement (e.g. if using online mode and a schema change has occurred)
+		// then we get SQL statements like this:
+		// insert into "UNKNOWN"."OBJ# 45522"("COL 1","COL 2","COL 3","COL 4") values (HEXTORAW('45465f4748'),HEXTORAW('546563686e6963616c20577269746572'), HEXTORAW('c229'),HEXTORAW('c3020b'));
+		// we additionally get status=2 for these records. Status 2 means the SQL statement is not valid for redoing.
+		// Some versions of Oracle report STATUS=2 for LONG column types, but have an empty info column, whereas when
+		// status=2 and there is some reason for the error in the info column (usually "Dictionary Mismatch") we consider
+		// the case to be one of dictionary mismatch
+		if msg.status == 2 && msg.info != "" {
+			return nil, fmt.Errorf("dictionary mismatch (%s) for table %q: %q", msg.info, msg.tableName, msg.sql)
+		}
+
+		msgs = append(msgs, msg)
+	}
+
+	if err := rows.Err(); err != nil {
+		return nil, err
+	}
+
+	return msgs, nil
+}
+
+func (s *replicationStream) tableActive(streamID string) bool {
+	s.tables.RLock()
+	defer s.tables.RUnlock()
+	var _, ok = s.tables.active[streamID]
+	return ok
+}
+
+func (s *replicationStream) keyColumns(streamID string) ([]string, bool) {
+	s.tables.RLock()
+	defer s.tables.RUnlock()
+	var keyColumns, ok = s.tables.keyColumns[streamID]
+	return keyColumns, ok
+}
+
+func (s *replicationStream) ActivateTable(ctx context.Context, streamID string, keyColumns []string, discovery *sqlcapture.DiscoveryInfo, metadataJSON json.RawMessage) error {
+	s.tables.Lock()
+	s.tables.active[streamID] = struct{}{}
+	s.tables.keyColumns[streamID] = keyColumns
+	s.tables.discovery[streamID] = discovery
+	s.tables.Unlock()
+	return nil
+}
+
+// Acknowledge informs the ReplicationStream that all messages up to the specified
+// SCN have been persisted
+func (s *replicationStream) Acknowledge(ctx context.Context, cursor string) error {
+	logrus.WithField("cursor", cursor).Debug("advancing acknowledged SCN")
```

Comment:
The `startCursor` is from the connector's persisted state (see [here](https://github.com/estuary/connectors/blob/051a0503d1f4023c6bda58a635684f185b6976b3/sqlcapture/capture.go#L163) for the invocation of `ReplicationStream`).

I believe the only current SQL capture that does anything with runtime acknowledgements is `source-postgres`, which [uses the runtime acknowledgements](https://github.com/estuary/connectors/blob/051a0503d1f4023c6bda58a635684f185b6976b3/source-postgres/replication.go#L664-L701) that indicate that the checkpoint (and its documents) has been durably committed to Flow to advance the postgres transaction log to a specific LSN, basically telling postgres that the connector will never need to return to any prior LSN, which is known since it has gotten confirmation from the runtime that the commit up until that LSN was durably committed. Other SQL captures (ex: [sqlserver](https://github.com/estuary/connectors/blob/051a0503d1f4023c6bda58a635684f185b6976b3/source-sqlserver/replication.go#L161-L163), [mysql](https://github.com/estuary/connectors/blob/051a0503d1f4023c6bda58a635684f185b6976b3/source-mysql/replication.go#L882-L890)) do not have anything in particular to do with runtime acknowledgements.

As the code is written here the `cursor` is being set to `s.ackSCN`, which appears to be similar to `source-postgres`, but then `s.ackSCN` is never accessed anywhere else. I would say at least as its written this could instead be a no-op like sqlserver/mysql/etc., unless there is actually something oracle should be doing in response to runtime acknowledgements, like informing the server that it can delete files or something. If it's possible or necessary to do that I don't see it happening anywhere in the code yet.